polar-core/src/types.rs

```diff
@@ -588,8 +590,16 @@ impl KnowledgeBase {
     }
 
     /// Return a monotonically increasing integer ID.
+    ///
+    /// Wraps around at 52 bits of precision so that it can be safely coerced to an IEEE-754
+    /// double-float (f64).
     pub fn new_id(&self) -> u64 {
-        self.id_counter.fetch_add(1, Ordering::SeqCst)
+        if self.id_counter.load(Ordering::SeqCst) == MAX_ID {
```

Comment:
If you do the check after the fetch, you save an operation on most calls.

Reply:
Actually, might need to do https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html#method.compare_exchange instead. Otherwise, there's a weird race condition where two threads might both set the counter to 1.
(load -> load -> store -> store).

Reply:
Yeah I think you're right. Changing now.